sql/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java

```diff
@@ -137,13 +139,95 @@ public AnalyzedRelation analyze(Node node, Analysis analysis) {
+                    relationAnalysisContext.sources(),
+                    relationAnalysisContext.parentSources(),
+                    statementContext.transactionContext().sessionContext().defaultSchema()),
+                new SubqueryAnalyzer(this, statementContext));
+            ExpressionAnalysisContext expressionAnalysisContext = relationAnalysisContext.expressionAnalysisContext();
+            SelectAnalysis selectAnalysis = new SelectAnalysis(
+                childRelationFields.size(),
+                relationAnalysisContext.sources(),
+                expressionAnalyzer,
+                expressionAnalysisContext);
+            for (Field field : childRelationFields) {
+                selectAnalysis.add(field.path(), field);
+            }
+
+            return new OrderedLimitedRelation(
+                    childRelation,
+                    analyzeOrderBy(
+                        selectAnalysis,
+                        node.getOrderBy(),
+                        expressionAnalyzer,
+                        expressionAnalysisContext,
+                        false,
+                        false),
+                    longSymbolOrNull(node.getLimit(), expressionAnalyzer, expressionAnalysisContext),
+                    longSymbolOrNull(node.getOffset(), expressionAnalyzer, expressionAnalysisContext));
+        }
+        return process(node.getQueryBody(), statementContext);
     }
 
     @Override
     protected AnalyzedRelation visitUnion(Union node, StatementAnalysisContext context) {
-        throw new UnsupportedFeatureException("UNION is not supported");
+        if (node.isDistinct()) {
+            throw new UnsupportedFeatureException("UNION [DISTINCT] is not supported");
+        }
+        QueriedRelation left = (QueriedRelation) process(node.getLeft(), context);
+        QueriedRelation right = (QueriedRelation) process(node.getRight(), context);
+
+        ensureUnionOutputsHaveTheSameSize(left, right);
+        ensureUnionOutputsHaveCompatibleTypes(left, right);
+
+        UnionSelect unionSelect = new UnionSelect(left, right, node.isDistinct());
+        unionSelect.querySpec().outputs(left.querySpec().outputs());
+        return unionSelect;
+    }
+
+    private static void ensureUnionOutputsHaveTheSameSize(QueriedRelation left, QueriedRelation right) {
+        if (left.outputs().size() != right.outputs().size()) {
+            throw new UnsupportedOperationException("Number of output columns must be the same for all parts of a UNION");
+        }
+    }
+
+    private static void ensureUnionOutputsHaveCompatibleTypes(QueriedRelation left, QueriedRelation right) {
+        List<Symbol> leftOutputs = left.outputs();
+        List<Symbol> rightOutputs = right.outputs();
+        for (int i = 0; i < leftOutputs.size(); i++) {
+            Symbol leftSymbol = leftOutputs.get(i);
+            Symbol rightSymbol = rightOutputs.get(i);
+            Symbol result = Symbols.tryUpcast(leftSymbol, rightSymbol);
```

Comment:
Does it make sense to also try to cast to the right symbol? I'd expect it to have the first relation enforce the types and have the second relation coerced to them. But not the other way around.

Reply:
in Postgres:
```
select col1 from t1
union all
select col1::bigint from t2;
```

gives a `bigint` column as result